src/System.Security.Cryptography.X509Certificates/src/Internal/Cryptography/Pal.OSX/ChainPal.cs

```diff
@@ -0,0 +1,489 @@
+            X509Certificate2Collection extraStore,
+            bool checkRevocation)
+        {
+            SafeCreateHandle policiesArray = PreparePoliciesArray(checkRevocation);
+            SafeCreateHandle certsArray = PrepareCertsArray(leafCert, extraStore);
+
+            int osStatus;
+
+            SafeX509ChainHandle chain;
+            int ret = Interop.AppleCrypto.AppleCryptoNative_X509ChainCreate(
+                certsArray,
+                policiesArray,
+                out chain,
+                out osStatus);
+
+            if (ret == 1)
+            {
+                _chainHandle = chain;
+                return;
+            }
+
+            chain.Dispose();
+
+            if (ret == 0)
+            {
+                throw Interop.AppleCrypto.CreateExceptionForOSStatus(osStatus);
+            }
+
+            Debug.Fail($"AppleCryptoNative_X509ChainCreate returned unexpected return value {ret}");
+            throw new CryptographicException();
+        }
+
+        public void Dispose()
+        {
+            if (_extraHandles == null)
+                return;
+
+            Stack<SafeHandle> extraHandles = _extraHandles;
+            _extraHandles = null;
+
+            _chainHandle?.Dispose();
+
+            while (extraHandles.Count > 0)
+            {
+                extraHandles.Pop().Dispose();
+            }
+        }
+
+        public bool? Verify(X509VerificationFlags flags, out Exception exception)
+        {
+            exception = null;
+
+            return ChainVerifier.Verify(ChainElements, flags);
+        }
+
+        private SafeCreateHandle PreparePoliciesArray(bool checkRevocation)
+        {
+            IntPtr[] policies = new IntPtr[checkRevocation ? 2 : 1];
+
+            SafeHandle defaultPolicy = Interop.AppleCrypto.X509ChainCreateDefaultPolicy();
+
+            if (defaultPolicy.IsInvalid)
+            {
+                defaultPolicy.Dispose();
+                throw new PlatformNotSupportedException(nameof(X509Chain));
+            }
+
+            _extraHandles.Push(defaultPolicy);
+            policies[0] = defaultPolicy.DangerousGetHandle();
+
+            if (checkRevocation)
+            {
+                SafeHandle revPolicy = Interop.AppleCrypto.X509ChainCreateRevocationPolicy();
+                _extraHandles.Push(revPolicy);
+                policies[1] = revPolicy.DangerousGetHandle();
+            }
+
+            SafeCreateHandle policiesArray =
+                Interop.CoreFoundation.CFArrayCreate(policies, (UIntPtr)policies.Length);
+
+            _extraHandles.Push(policiesArray);
+            return policiesArray;
+        }
+
+        private SafeCreateHandle PrepareCertsArray(ICertificatePal cert, X509Certificate2Collection extraStore)
+        {
+            IntPtr[] ptrs = new IntPtr[1 + (extraStore?.Count ?? 0)];
+
+            AppleCertificatePal applePal = (AppleCertificatePal)cert;
+
+            ptrs[0] = applePal.CertificateHandle.DangerousGetHandle();
```

Comment:
My standard question/comment every time I see DangerousGetHandle: let's just make sure we've audited all of these to ensure we've either not yet given out the SafeHandle so we know it couldn't have been disposed, or it's been appropriately addref'd.